### PR TITLE
Fix reasoning prefill auto parse

### PR DESCRIPTION
**Pull request by Reithan**
This PR fixes two bugs in the reasoning system when backends provide native `reasoning_content` fields:

## Auto-parse Content Dropping Bug

**Problem:** When backends return identical `reasoning_content` and `content`, SillyTavern reconstructs `<think>` tags then auto-parses them again, causing message content to disappear.

**Fix:** Skip auto-parsing when reasoning comes from native API (`ReasoningType.Model`).

**Files:** `public/scripts/reasoning.js`

## Prefill Contamination Bug  

**Problem:** Reasoning prefills (like `"<think>\n"`) appear in message content instead of reasoning.

**Fix:** Unified reasoning prefix handling that separates reasoning from content correctly.

**Files:** `public/script.js`

## Additional Edge Case Fixes

**Type Safety & Null Handling:**
- Fixed nullable type annotations for DOM elements and timing properties that can be null
- Added null safety for message content parsing and duration handling
- Prevented undefined values from being stored in reasoning extras

**Enhanced Prefill Processing:**
- Improved `user_prompt_bias` handling for native model reasoning to prevent double-inclusion
- Added logic to strip reasoning prefixes that leak into message content  
- Enhanced re-parsing when reasoning contains both prefix and suffix markers
- Added precise control over bias inclusion with `includeUserPromptBias` flag

**Reasoning State Management:**
- Better state transitions for `ReasoningType.Model` when reasoning is provided natively
- Improved reasoning state handling to prevent incorrect state persistence

**Files:** `public/scripts/reasoning.js`, `public/script.js`

Both the core fixes and edge case handling ensure native reasoning APIs work without content corruption, contamination, or state management issues.

## Tests Run

| Prefill | Streaming | Non-Streaming | Continue Streaming | Continue Non-Streaming |
| --- | :---: | :---: | :---: | :---: |
| `null` | ✓ | ✓ | ✓ | ✓ |
| `{{prefix}}` | ✓ | ✓ | ✓ | ✓ |
| `{{prefix}}\n` | ✓ | ✓ | ✓ | ✓ |
| `{{prefix}}\nOk, so the user` | ✓ | ✓ | reasoning: finished:✓/continuing:✓ | ✓ |

## Checklist:

- [X] I have read the [Contribution guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Reithan** [comment]
Found a couple more edge cases, will un-draft when completed.

**Comment by Reithan** [comment]
The core issue behind most of the edge cases is that different types of prefills affect the output of backend differently, making it swap from 'model' to 'parsed' type reasoning, sometimes returning blank reasoning (immediately closed), etc.

**Comment by Reithan** [comment]
Ok, going back to draft. Found a bug in continue, but all the non-continue cases are working.

**Comment by Reithan** [comment]
Fixed

**Comment by Reithan** [comment]
Added breakdown of tests run locally in PR summary

**Comment by Reithan** [comment]
Found another edge-of-an-edge in streaming continues with longer prefill. Addressing.

**Comment by Reithan** [comment]
One more checkmark for the checkmark throne.

**Comment by Reithan** [comment]
Few more hours of testing, no other breaks. Should be solid, now. 👍

File: public/scripts/reasoning.js [hunk 1]
```diff
@@ -273,9 +273,9 @@ export class ReasoningHandler {
         this.reasoning = '';
         /** @type {string?} The reasoning output display in case of translate or other */
         this.reasoningDisplayText = null;
-        /** @type {Date} When the reasoning started */
+        /** @type {Date?} When the reasoning started */
```
Comment (Wolfsblvt) [inline public/scripts/reasoning.js]: Usually, `?` is used for optional things, like function parameters that don't need to be set or properties on an object that don't need to be initialized, etc.

In places like this, the better writing would be `Date|null` to explicitly say this value can be set to `null` manually.

File: public/scripts/reasoning.js [hunk 2]
```diff
@@ -396,24 +403,26 @@ export class ReasoningHandler {
...
+        this.type = (this.#isParsingReasoning || this.#parsingReasoningMesStartIndex) ? ReasoningType.Parsed : reasoning ? ReasoningType.Model : null;
+
+        if (reasoningChanged && this.type == ReasoningType.Model && this.state == ReasoningState.None)
+            this.state = this.reasoning ? ReasoningState.Thinking : ReasoningState.Done;
 
         if (persist) {
             // Build and save the reasoning data to message extras
             extra.reasoning = this.reasoning;
-            extra.reasoning_duration = this.getDuration();
+            extra.reasoning_duration = this.getDuration() ?? undefined;
```
Comment (Wolfsblvt) [inline public/scripts/reasoning.js]: What's the reason to unset it to `undefined` explicitly?
`getDuration()` already returns `null`. And I feel like once a reasoning is set, all reasoning properties should exist. So explictly using null works better.

File: public/script.js [hunk 3]
```diff
@@ -5433,24 +5461,38 @@ export async function Generate(type, { automatic_trigger, force_name2, quiet_pro
...
 
         if (power_user.trim_spaces) {
             reasoning = reasoning.trim();
         }
 
-        if (isContinue) {
-            continue_mag = promptReasoning.removePrefix(continue_mag);
-            getMessage = continue_mag + getMessage;
-        }
+        messageChunk = cleanUpMessage({
```
Comment (Wolfsblvt) [inline public/script.js:5522]: You are sure we can just switch this over to a full cleanup?
Cleanup also does run regexes, substitutes markdown, etc etc..
That's a big change here.

**Review by Wolfsblvt — COMMENTED** [review]
Didn't have the time to test it thoroughly with reasoning providers. But I have some short feedback already.

File: public/scripts/reasoning.js (hunk 2 above)
Comment (Reithan) [reply public/scripts/reasoning.js]: The type hints define duration as allowing undefined but not null, so this was showing a warning. Was just a minor nit fix. Not important, I can drop if you mind it.

File: public/script.js (hunk 3 above)
Comment (Reithan) [reply public/script.js:5522]: This isn't changed, the steps here are just re-ordered slightly. This cleanup was already here.

File: public/scripts/reasoning.js (hunk 2 above)
Comment (Wolfsblvt) [reply public/scripts/reasoning.js]: Maybe we should correct the jsdoc then allowing null. Because this is how it worked so far, and it still makes sense to me.

**Comment by Reithan** [comment]
@Wolfsblvt Is there any specific testing or validation I can do around this to help?

**Comment by Wolfsblvt** [comment]
Not really. We just need to find the time to test this thoroughly, as its not a very easy test scenario.

**Comment by lukasstraub2** [comment]
> **Problem:** When backends return identical `reasoning_content` and `content`, SillyTavern reconstructs `<think>` tags then auto-parses them again, causing message content to disappear.

With what backend/provider/inference engine does this actually happen? I was not able to reproduce this.

**Comment by lukasstraub2** [comment]
> **Problem:** Reasoning prefills (like `"<think>\n"`) appear in message content instead of reasoning.

I can't reproduce this either, again under what conditions does this actually happen?

**Comment by Reithan** [comment]
All my testing has been against locally hosted models using KoboldCpp.

I'll update to latest versions and test again, soon.

**Comment by Reithan** [comment]
Convert to draft until I can retest

**Comment by Reithan** [comment]
Ok, running a new set of tests directly against `staging`. This is all in CC mode against local koboldcpp instance.

## Tests Run (against `staging`)
| Prefill | Streaming | Non-Streaming | Continue Streaming<br>[post-reasoning] | Continue Streaming<br>[mid-reasoning] | Continue Non-Streaming<br>[post-reasoning] | Continue Non-Streaming<br>[mid-reasoning] |
| --- | :---: | :---: | :---: | :---: | :---: | :---: |
| `null` | ✓ | ✓ | ✓ | ✓ | ✓ | **✗** |
| `{{prefix}}` | **✗** | **✗** | ✓ | ✓ | ✓ | **✗** |
| `{{prefix}}\n` | **✗** | **✗** | ✓ | ✓ | ✓ | **✗** |
| `{{prefix}}\nOk, so the user` | ✓ | **✗** | ✓ | ✓ | ✓ | **✗** |

In all cases, the failed cases look like the prefill being re-appended to the beginning of the message after reasoning. Example:

image

Then I've re-run the same tests against a branch with my changes

## Tests Run (against with my changes)
| Prefill | Streaming | Non-Streaming | Continue Streaming<br>[post-reasoning] | Continue Streaming<br>[mid-reasoning] | Continue Non-Streaming<br>[post-reasoning] | Continue Non-Streaming<br>[mid-reasoning] |
| --- | :---: | :---: | :---: | :---: | :---: | :---: |
| `null` | ✓ | ✓ | ✓ | ✓ | ✓ | **✗** |
| `{{prefix}}` | ✓ | ✓ | ✓ | ✓ | ✓ | **✗** |
| `{{prefix}}\n` | ✓ | ✓ | ✓ | ✓ | ✓ | **✗** |
| `{{prefix}}\nOk, so the user` | ✓ | ✓ | ✓ | ✓ | ✓ | **✗** |

So, it looks like I still have that last edge case to fix: non-streaming mid-reasoning continue/resume.

**Comment by Reithan** [comment]
Ok, new commit closes the last edge case. This should be ready for review again.
